### DISCUSSION TITLE
QUESTION ABOUT TRAINING TEXT DETECTION MODEL

### DISCUSSION THREAD

**Discussion by anhalu** [mindee/doctr#1386] [Q&A] [answered]
when I train a text detection model using pretrained db_resnet50, I find it quite slow to calculate. it only use 1 thread in CPU, ( I have 24 thread in my CPU and 24GB in CUDA ). How do I speed up the calculation faster ?  
I use pytorch and use this command for training scrip : 
python references/detection/train_pytorch.py path/to/your/train_set path/to/your/val_set db_resnet50 --epochs 5 --device 0 --pretrained 
I see at base.py in DB of doctr model, it has a lot of numpy, so it is using too much CPU to calculate, how can I increase the thread CPU for training or use more GPU for training ? I need help. Thank a lots

**Comment by anhalu** [comment]
@felixdittrich92 can u help me ? thank u so much

**Comment by anhalu** [comment]
@odulcy-mindee  plz, can you take a look ? thank a lot :((

**Comment by felixT2K** [comment]
Hi @anhalu 👋🏼,

Excuse the late response 😅 
`1 thread in CPU` this sounds weird ... could you share some more information about this behaviour (maybe `htop` log)

Try to specify the workers by yourself with `-j <NUMBER_WORKERS>` (e.g.: -j 16) and with 24 GB VRAM you can also increase the batch size (try 12 / 14 / 16) 

This has nothing to do with the models postprocessing it's more we apply by default a lot of augmentations which takes some time. 
Feel free to modify this to your needs :)

https://github.com/mindee/doctr/blob/8e0609d3e3618750ce0915c8f1ab6f3c6c5e64db/references/detection/train_pytorch.py#L269

**Comment by anhalu** [comment]
![image](https://github.com/mindee/doctr/assets/89495585/b8712c92-b355-4a8f-857c-493d0bd72bf1)

**Comment by anhalu** [comment]
@felixT2K  when training, i see in htop, it only use 1 thread cpu to calculator, and it takes up 5gb gpu, I don't think it's a data loading problem, I think it has a problem with the loss calculation of base.py in detection forder. I see a lot of places using numpy in base.py, does that affect the calculation when training the model? and how do I speed up the training process

**Comment by anhalu** [comment]
![image](https://github.com/mindee/doctr/assets/89495585/8654cffc-0fb2-42fa-9c21-b328c77e27cd)
My data is only 27 sample to test training, but It takes a lot of time to process calculations on the CPU side. Did I do anything wrong at any stage? Thanks a lot for response

**Reply by odulcy-mindee** [reply]
@anhalu Do you see any activity on GPU ?

As @felixT2K said, it can take some time to start due to the augmentation pipeline

**Reply by anhalu** [reply]
@odulcy-mindee  It runs on the gpu, the gpu runs quite fast, then goes to the cpu, when the cpu runs out it runs again to the gpu, but the processing on the cpu takes a lot of time while there are only 27 samples. plz, help me

**Reply by odulcy-mindee** [reply]
@anhalu Can you tell us more about your environment please (doctr version, python version).
It looks like you are using `cuda >12`. Maybe you can give a try to a Docker Image we got (which is Cuda 11), it may helps.

**Reply by felixdittrich92** [reply]
https://github.com/mindee/doctr/pkgs/container/doctr

**Reply by anhalu** [reply]
![image](https://github.com/mindee/doctr/assets/89495585/e70c781e-4a65-40d3-84f5-c0d53cfd3599)
![image](https://github.com/mindee/doctr/assets/89495585/c34c79bf-cf95-4b9e-829e-f320695638a1)
![image](https://github.com/mindee/doctr/assets/89495585/175f55fd-c7a5-4441-8482-853d4cd7eab2)

so for cuda > 12, doctr cannot train properly ? I do the samething with colab, but it slow too (the third image is from colab). This is so sad. I will try  a Docker Image for training process

**Reply by felixT2K** [reply]
@anhalu please try to uncomment every augmentation expect the resizing

The code works we have running the same code atm we need only to find the issue on your side 😅

**Reply by anhalu** [reply]
@felixT2K  I don't do anything in lib, I just git clone and run the command " python references/detection/train_pytorch.py /home/anhm/data/word_det/data_doctr /home/anhm/data/word_det/data_doctr db_resnet50 --epochs 50 --device 0 --pretrained -j 16
". and it run slowly. it took 16 minute to finish one epochs.

**Reply by anhalu** [reply]
![image](https://github.com/mindee/doctr/assets/89495585/87e998cb-f4cd-4dbd-85d3-ea1fe0696951)

@felixT2K  oh, I have question, Does it matter if I leave img_hash as empty in labels.json when I prepare the data for training ?

**Reply by felixT2K** [reply]
No that's fine

**Reply by felixT2K** [reply, marked as answer]
but .. your polygon annotations looks not correct 

-> 'polygons': [[[x1, y1], [x2, y2], [x3, y3], [x4, y4]], ...]

**Reply by anhalu** [reply]
Oh, It need 4 point ?, I just use top left and bottom right for rectange box. let's me try prepare data again. Does it need to be calculated in any order?

**Reply by felixT2K** [reply]
```
The order of the points does not matter inside a polygon. Points are (x, y) absolutes coordinates.
```

:)

**Reply by anhalu** [reply]
Thank a lot, it worked. I didn't carefully check the number of waypoints. <3

**Reply by anhalu** [reply]
I missed 2 point, so it cannot run property. I just fix that, So it worked now.

**Reply by felixT2K** [reply]
great 🤗
